SQLite, MySQL, MariaDB, Microsoft SQL Server providers support and `@loancrate/prisma-schema-parser` integration

Jakoma02 opened the pull request:
This PR provides support for the SQLite database and provides an easy way to support other databases in the future.

I implemented this feature to experiment with it in my personal project; hopefully, it will be useful to others as well. I briefly tested the functionality with both SQLite and Postgres, but the testing was by no means extensive, and I will be testing it more as I go.

Thank you for the great project, @Softjey! It definitely still has a long way to go, but it has helped me solve my problem with data migrations :slightly_smiling_face: 

## Included Changes

- The `slonik` library was replaced with the `knex.js` library, which allows building and running SQL queries in a database-agnostic way.
- The `datasource` block from the Prisma schema is now used to determine which database to use
  - Using environment variables (e.g. `url = env(DATABASE_URL)`) is supported
  - Relative paths are correctly resolved relative to the main Prisma schema
- Modification to the schema for client generations was reworked:
  - For parsing the schema, the `@loancrate/prisma-schema-parser` library is now used for parsing and formatting of the schema to avoid fragile manual parsing of the schema
  - All generator blocks other than the JavaScript client generator are removed from the schema to avoid generating unrelated clients at wrong relative paths (this is not directly related to SQLite, but it made sense to address this together with the extensive changes in schema modification)
  - If the `datasource` block uses the `sqlite` provider, its URL is resolved from an environment variable (if it is used), and if the path to the database is relative, it is correctly resolved relative to the main Prisma schema.
  - The modifications are no longer made in-place in the copied schema, but are written to a schema in the temporary directory directly prior to generating the client. The reason for this is that absolute paths differ between machines, so it is not wise to keep these modifications in source control.
- Functions related to work with the temporary directory were extracted to a `tempDir.js` util file to allow reuse. A `withTempDir` helper function was added to ensure that the temporary directory is removed in case of an error.

Jakoma02 committed:
SQLite support

Softjey commented:
Hi, thank you for your contribution! I’ll review the changes in the next few days.

Softjey committed:
refactor(tempMigrationSchema): format code and replace comments by js docs

Softjey committed:
refactor(tempDir): format code and improve JSDoc comments

Softjey committed:
feat: add isSupportedDatasourceProvider utility

Softjey committed:
refactor(readDataSourceConfig): move prismaSqliteURLToFilePath util to separate folder, add supported providers validation, replace comments by js docs

Softjey committed:
refactor(DB): update DataSourceConfig to use SupportedDatasourceProvider type, format code

Softjey committed:
refactor(CLI): convert generate method to async and optimize type generation with Promise.all

Softjey committed:
refactor(cli): rename prisma variable to db for clarity and convert generate action to async

Softjey committed:
fix(tempMigrationSchema): create output attribute in schema if it's not exist, rewrite reading and writing files to fs promises

Softjey committed:
fix(CLI): update generate method to await createTempSchema and log completion message instead of starting message

Softjey committed:
fix(TargetedPrismaMigrator): await withTempDir to ensure proper execution of migration file handling

Softjey committed:
fix(readDataSourceConfig): include supported datasource providers in error message for unsupported provider

Softjey committed:
feat: add support for MySQL and SQL Server in database configuration

- Updated package.json to include MySQL and tedious (SQL Server) dependencies.
- Enhanced createKnexConfig function to support MySQL and SQL Server connections.
- Modified isSupportedDatasourceProvider to include MySQL and SQL Server as supported providers.

Softjey committed:
docs: update README to include Database Support section and clarify supported databases

Softjey committed:
fix(DB): await knex query in getMigrationByName to ensure proper execution

Softjey committed:
style: format SUPPORTED_DATASOURCE_PROVIDERS array for improved readability

Softjey committed:
fix(readArgumentWithEnv): add error handling for unset environment variables

Softjey commented:
Hi! Thanks for adding the foundation for multi‑provider support and the ready Prisma schema parser. That’s a significant contribution to the library’s growth. I’ve done a bit of refactoring, fixed the bugs I found, and added support for a few more SQL providers. After we merge this, I’ll publish a new version to npm.

Could you please confirm everything works for you with SQLite? If it does, feel free to open a small PR updating the README.md in the support table, mark SQLite as tested.

Thanks again for your contribution!

Jakoma02 commented:
Hi, thanks for cleaning up the code and adding the other providers, it's great seeing the changes merged in so quickly :slightly_smiling_face: 

During more thorough testing, I realized that I missed something that seems rather obvious in hindsight: the environment variables used in the schema file are often not available at build time when the generate command is run, but only at runtime. Because we are now substituting the environment variable values at generation time, changes at runtime are not picked up.

I managed to resolve this by changing the approach to environment variables a bit: instead of substituting them, I leave them in the schema. That gives back the issue with relative paths resolved relative to the new schema instead of the original one, but it turns out that there is a simple solution: instead of creating the temporary schema in a temporary directory, it can be created directly alongside the original schema. That way, paths relative to both files will be resolved the same way, and even in the case of relative paths, the environment variable value will be correctly resolved by the client at runtime.

I have the changes mostly ready. I will rebase them to your cleaned-up version, do some more testing, and create a PR. In the meantime, I'd be interested to hear your thoughts if you have any!
